src/perf/event.cpp

```diff
@@ -0,0 +1,658 @@
+
+extern "C"
+{
+#include <fcntl.h>
+#include <sys/ioctl.h>
+}
+
+namespace lo2s
+{
+namespace perf
+{
+
+constexpr std::uint64_t operator"" _u64(unsigned long long int lit)
+{
+    return static_cast<std::uint64_t>(lit);
+}
+
+constexpr std::uint64_t bit(int bitnumber)
+{
+    return static_cast<std::uint64_t>(1_u64 << bitnumber);
+}
+
+// helper for visit function
+template <class... Ts>
+struct overloaded : Ts...
+{
+    using Ts::operator()...;
+};
+template <class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+
+template <typename T>
+T read_file_or_else(std::string filename, T or_else)
+{
+    T val;
+    std::ifstream stream(filename);
+    stream >> val;
+    if (stream.fail())
+    {
+        return or_else;
+    }
+    return val;
+}
+
+static std::uint64_t parse_bitmask(const std::string& format)
+{
+    enum BITMASK_REGEX_GROUPS
+    {
+        BM_WHOLE_MATCH,
+        BM_BEGIN,
+        BM_END,
+    };
+
+    std::uint64_t mask = 0x0;
+
+    static const std::regex bit_mask_regex(R"((\d+)?(?:-(\d+)))");
+    const std::sregex_iterator end;
+    std::smatch bit_mask_match;
+    for (std::sregex_iterator i = { format.begin(), format.end(), bit_mask_regex }; i != end; ++i)
+    {
+        const auto& match = *i;
+        int start = std::stoi(match[BM_BEGIN]);
+        int end = (match[BM_END].length() == 0) ? start : std::stoi(match[BM_END]);
+
+        const auto len = (end + 1) - start;
+        if (start < 0 || end > 63 || len > 64)
+        {
+            throw EventProvider::InvalidEvent("invalid config mask");
+        }
+
+        /* Set `len` bits and shift them to where they should start.
+         * 4-bit example: format "1-3" produces mask 0b1110.
+         *    start := 1, end := 3
+         *    len  := 3 + 1 - 1 = 3
+         *    bits := bit(3) - 1 = 0b1000 - 1 = 0b0111
+         *    mask := 0b0111 << 1 = 0b1110
+         * */
+
+        // Shifting by 64 bits causes undefined behaviour, so in this case set
+        // all bits by assigning the maximum possible value for std::uint64_t.
+        const std::uint64_t bits =
+            (len == 64) ? std::numeric_limits<std::uint64_t>::max() : bit(len) - 1;
+
+        mask |= bits << start;
+    }
+    Log::debug() << std::showbase << std::hex << "config mask: " << format << " = " << mask
+                 << std::dec << std::noshowbase;
+    return mask;
+}
+
+static constexpr std::uint64_t apply_mask(std::uint64_t value, std::uint64_t mask)
+{
+    std::uint64_t res = 0;
+    for (int mask_bit = 0, value_bit = 0; mask_bit < 64; mask_bit++)
+    {
+        if (mask & bit(mask_bit))
+        {
+            res |= ((value >> value_bit) & bit(0)) << mask_bit;
+            value_bit++;
+        }
+    }
+    return res;
+}
+
+Event::Event() : name_("")
+{
+    memset(&attr_, 0, sizeof(attr_));
+    attr_.size = sizeof(attr_);
+    attr_.type = PERF_TYPE_RAW;
+
+    attr_.config = 0;
+    attr_.config1 = 0;
+
+    attr_.sample_period = 0;
+    attr_.exclude_kernel = 1;
+}
```

Comment:
I don't think a default constructed event should be possible.

Reply:
Why not? This way you can say that a reader holds an event without needing to know the specifics.

Reply:
Because of the unknown specifics, you end up with an object that cannot be used. It's not initialized, a constructor should never give an ususable object.